### PR TITLE
Travis build proof of concept

### PR DESCRIPTION
**Pull request by giorgiosironi**
See README for description.

unstable-environment-process
(generated from https://www.draw.io/#G1qKnrenkxHFJaXt3D1ZzqDO1gxAl1sc1X not available for sharing at this time)

### CODE REVIEW

**Comment by giorgiosironi** [comment]
`passed` because it does not deploy on pull requests, of course (it's an empty build).

File: scripts/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+set -e
+
+if [ "$#" -ne 1 ]; then
+    echo "USAGE $0 ENVIRONMENT_NAME"
+    exit 1
+fi
+
+environment_name="$1"
+hostname="${environment_name}.libero.pub"
+ssh "$hostname" /usr/local/bin/deploy browser="$(scripts/latest-revision.sh git@github.com:libero/browser)"
```
Comment (giorgiosironi) [inline scripts/deploy.sh]: the simplest form `/usr/local/bin/deploy` could take is
```
git clone git@github.com:libero/my-sample-configuration
cd my-sample-configuration
./update.sh $@ # updates .env with passed versions, runs `docker-compose up`
```
I can inline this here to be more opinionated. Opinions?

Comment (giorgiosironi) [reply scripts/deploy.sh]: (meaning: executing this steps through SSH)

File: scripts/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+if [ "$#" -ne 1 ]; then
+    echo "Usage: $0 ENVIRONMENT_NAME"
+    exit 1
+fi
+
+environment_name="$1"
+hostname="${environment_name}.libero.pub"
+scp scripts/deploy.sh "$hostname":/tmp/remote-deploy.sh
+ssh "$hostname" REVISION_BROWSER="$(scripts/latest-revision.sh git@github.com:libero/browser)" /tmp/remote-deploy.sh
```
Comment (lsh-0) [inline scripts/deploy.sh]: > REVISION_BROWSER="$(scripts/latest-revision.sh git@github.com:libero/browser)"

Is this running locally or remotely? I would say locally and the result becomes part of the command ssh runs.

File: scripts/latest-revision.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# with minimal cloning
+set -e
+
+if [ "$#" -ne 1 ]; then
+    echo "USAGE $0 GIT_REMOTE_REPOSITORY"
+    exit 1
+fi
+
+remote="$1"
+folder=$(basename "$remote")
```
Comment (lsh-0) [inline scripts/latest-revision.sh:12]: I'm not suggesting you change it, but `basename \/` => `/` which then becomes `rm -rf "/"` two lines later.

**Comment by lsh-0** [comment]
so, is 'unstable' in the graph the remote vm? if so, then does this read true?:

* infrastructure creates the remote vm
* this ('environments') repo can ssh into and create/update the configuration of the remote project
* the remote project pulls down docker images and it's configuration
* remote project launches containers described in it's configuration

and then:

* build tools like Travis can affect the project configuration and Docker images to merge passing builds
* developers can affect the unstable branch of the project
* developer changes trigger build/deployment tools, build/deployment tools trigger deploys

**Review by lsh-0 — APPROVED** [review]
it's nice and neat. I look forward to seeing it grow

File: scripts/deploy.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+set -e
+
+if [ "$#" -ne 1 ]; then
+    echo "Usage: $0 ENVIRONMENT_NAME"
+    exit 1
+fi
+
+environment_name="$1"
+hostname="${environment_name}.libero.pub"
+scp scripts/deploy.sh "$hostname":/tmp/remote-deploy.sh
```
Comment (giorgiosironi) [inline scripts/deploy.sh]: bug: this should be `scripts/remote-deploy.sh`

File: scripts/deploy.sh (hunk 2 above)
Comment (giorgiosironi) [reply scripts/deploy.sh]: Todo: extract intermediate variable for `$revision_browser`

File: scripts/latest-revision.sh (hunk 3 above)
Comment (giorgiosironi) [reply scripts/latest-revision.sh:12]: Let's do some check here to avoid corner cases

**Comment by giorgiosironi** [comment]
> so, is 'unstable' in the graph the remote vm? if so, then does this read true?:

- [x] infrastructure creates the remote vm
- [x] this ('environments') repo can ssh into and create/update the configuration of the remote project
- [x] the remote project pulls down docker images and it's configuration
- [x] remote project launches containers described in it's configuration

(all true)

> build tools like Travis can affect the project configuration and Docker images to merge passing builds

They will pass in new tags for these Docker images, rolling out a new version.

> developers can affect the unstable branch of the project

If they merge to master, they affect the `unstable` environment (but it has no specific branch)

> developer changes trigger build/deployment tools, build/deployment tools trigger deploys

Yes, Github triggers Travis CI which triggers this which triggers `sample-configuration`.

**Pull request merged**